### PR TITLE
HV-1551 @Size annotation fails on not parametrized Map property

### PR DESCRIPTION
**Pull request by marko-bekhta**
- https://hibernate.atlassian.net/browse/HV-1551

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/size/SizeValidatorForCollection.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
  *
  * @author Hardy Ferentschik
  */
-public class SizeValidatorForCollection implements ConstraintValidator<Size, Collection<?>> {
+public class SizeValidatorForCollection implements ConstraintValidator<Size, Collection> {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/size/SizeValidatorForCollection.java]: Having a `SuppressWarnings` and a comment explaining why we did it like that would be nice.

It's not possible to keep the wildcard and fix it in another place?

File: engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/size/SizeValidatorForMap.java [hunk 2]
```diff
@@ -20,7 +20,7 @@
  *
  * @author Hardy Ferentschik
  */
-public class SizeValidatorForMap implements ConstraintValidator<Size, Map<?, ?>> {
+public class SizeValidatorForMap implements ConstraintValidator<Size, Map> {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/bv/size/SizeValidatorForMap.java]: Same here about the `SuppressWarnings` and comment if we can't fix it another way.

File: engine/src/test/java/org/hibernate/validator/test/constraints/annotations/SizeOnParametrizedAndNonParametrizedContainersTest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+import org.hibernate.validator.testutil.TestForIssue;
+
+import org.testng.annotations.Test;
+
+/**
+ * @author Marko Bekhta
+ */
+@TestForIssue(jiraKey = "HV-1551")
+public class SizeOnParametrizedAndNonParametrizedContainersTest extends AbstractConstrainedTest {
```
Comment (gsmet) [inline engine/src/test/java/org/hibernate/validator/test/constraints/annotations/SizeOnParametrizedAndNonParametrizedContainersTest.java]: Paramet*e*rized * 2 here.

And the same in the method names.

**Comment by gsmet** [comment]
I think it would also be interesting to check if it's a regression. And if so, what caused it exactly.

**Comment by marko-bekhta** [comment]
@gsmet I checked this against 5.4 and there's same issue. 
The problem is that when we are filtering through available validators `TypeHelper#isAssignable()` receives as a `supertype` parametrized type (`Map<?, ?>`) and as a `type` - simple `Map`. So in the end it calls `TypeHelper#isSuperAssignable()` where `getResolvedSuperclass( type )` is null and initial method returns `false`.

At first I thought about updating the `TypeHelper#isAssignable()` with something similar to:

```java
if ( supertype instanceof ParameterizedType ) {
// the first statement in this ^ `if`
    if ( type.equals( getErasedType( supertype ) ) ) {
        return true;
    }
}
```

But I wasn't 100% sure if it'll be a good idea to change this helper method as it's used in a couple more places... And also applying such change would make the next test fail
```java
public class TypeHelperTest {
    //.....    
    @Test
    public void isAssignableWithClassFromParameterizedType() {
        assertAsymmetricallyAssignable( List.class, TypeHelper.parameterizedType( List.class, Integer.class ) );
    }
    //.....
}
```

**Comment by gsmet** [comment]
@marko-bekhta I tried another approach, which I think is more correct (note that there are probably a lot of other ways to break `TypeHelper` but I think this will do for now).

All tests are passing.

**Comment by gsmet** [comment]
After a discussion on HipChat with Marko, I squashed the commits and merged. Thanks Marko for the initial work!

**Pull request merged**